index.js

```diff
@@ -1,91 +1,158 @@
+          }
+          else {
+              console.log(chalk.red(message));
+          }
+      });
   }
-  return this._optionsHash;
-};
-
-TypeScript.prototype.cacheKeyProcessString = function(string, relativePath) {
-  return this.optionsHash() + Filter.prototype.cacheKeyProcessString.call(this, string, relativePath);
-};
-
-TypeScript.prototype.processString = function (string, relativePath) {
-  try{
-    return ts.transpileModule(string, {compilerOptions: this.options, fileName: relativePath}).outputText;
-  }catch(e){
-    console.log("TYPESCRIPT ERROR: " + e.message)
-    console.log(e.stack + "\n");
+  
+   this.servicesHost = {
+      getScriptFileNames: function () { return self.rootFilenames; },
+      getScriptVersion: function (fileName) { return self.files[fileName] && self.files[fileName].version.toString(); },
+      getScriptSnapshot: function (fileName) {
+          if (!fs.existsSync(fileName)) {
+              return undefined;
+          }
+          return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString());
+      },
+      getCurrentDirectory: function () { return self.cwd; },
+      getCompilationSettings: function () {
+        return {
+          allowNonTsExtensions: false,
+          diagnostics: true,
+          noImplicitAny: true,
+          //outDir: self.outputPath, TODO: Get this working
+          module: 0,
+          preserveConstEnums: true,
+          removeComments: false,
+          sourceMap: false,
+          target: 2 // ES6
+        };
+      },
+      getDefaultLibFileName: function (options) { return ts.getDefaultLibFilePath(options); },
+  };
+  
+  this.services = ts.createLanguageService(this.servicesHost, ts.createDocumentRegistry());
+  
+  Plugin.call(this, inputNodes, {
+    persistentOutput: true
+  });
+}
 
-    return "";
+TypeScript.prototype.build = function() {
+  var self = this;
+  
+  self.inputPaths.forEach(function(srcDir) {
+    var paths = walkSync(srcDir, { directories: false });
+    paths.forEach(function (relativePath) {
+      if (!(relativePath in self.files)) {
+        // TODO: Handle removals
+        var absPath = path.join(srcDir, relativePath);
+        self.files[relativePath] = { 
+          absPath: absPath,
+          srcDir: srcDir,
+          version: 0,
+          stat: fs.statSync(absPath),
+          needsEmit: true 
+        };
+        self.rootFilenames.push(absPath);
+      }
+    });
+  });
+  
+  // Update version for modified files
+  self.inputPaths.forEach(function(srcDir) {
+    var paths = walkSync(srcDir, { directories: false });
+    paths.forEach(function (relativePath) {
+      var absPath = path.join(srcDir, relativePath);
+      var curr = fs.statSync(absPath);
+      var prev = self.files[relativePath].stat;
+      if (+curr.mtime <= +prev.mtime) {
+        return;
+      }
+      
+      self.files[relativePath].version += 1;
```

Comment:
you can use `curr` as a version if you want to.. it is treated as an opaque string on the ts side.